### PR TITLE
peerstream: dialer should reconnect when stream closes 

### PR DESCRIPTION
**Pull request by lkysow**
### Description
If the stream is closed unexpectedly (i.e. when we haven't received
a terminated message), the dialer should attempt to re-establish the
stream.

Previously, the `HandleStream` would return `nil` when the stream
was closed. The caller then assumed the stream was terminated on purpose
and so didn't reconnect when instead it was stopped unexpectedly and
the dialer should have attempted to reconnect.

### Testing & Reproduction steps
Start two servers, peer them, stop the accepting peer server and then restart it, see that the dialing server never reestablishes its connection.

### PR Checklist

* [x] updated test coverage
* [x] not a security concern

### CODE REVIEW

File: agent/consul/leader_peering_test.go [hunk 1]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"encoding/base64"
 	"encoding/json"
+	"github.com/hashicorp/consul/sdk/freeport"
```
Comment (acpana) [inline agent/consul/leader_peering_test.go:7]: nit: import ordering

File: agent/consul/leader_peering_test.go [hunk 2]
```diff
@@ -282,6 +283,120 @@ func TestLeader_Peering_DeferredDeletion(t *testing.T) {
...
+	// Re-insert the peering state.
+	require.NoError(t, acceptingServerRestart.fsm.State().PeeringWrite(2000, &pbpeering.Peering{
+		ID:    dialingServerPeerID,
+		Name:  "my-peer-dialing-server",
+		State: pbpeering.PeeringState_PENDING,
+	}))
+
+	// The dialing peer should eventually reconnect.
+	retry.Run(t, func(r *retry.R) {
+		connStreams := acceptingServerRestart.peeringServer.Tracker.ConnectedStreams()
```
Comment (acpana) [inline agent/consul/leader_peering_test.go:395]: nit:

```suggestion
		connStreams := acceptingServerRestart.peeringServer.ConnectedStreams()
```

File: agent/consul/leader_peering_test.go [hunk 3]
```diff
@@ -282,6 +283,120 @@ func TestLeader_Peering_DeferredDeletion(t *testing.T) {
...
+		require.Len(r, connStreams, 1)
```
Comment (acpana) [inline agent/consul/leader_peering_test.go:396]: nit: this is fine but could we actually look by the peer id of `dialingServerPeerID` and assert that it's in the map.

**Review by acpana — COMMENTED** [review]
this LGTM! 🚢 

Just a couple of non blocking nits

**Review by kisunji — APPROVED** [review]
LGTM!

File: agent/consul/leader_peering_test.go (hunk 1 above)
Comment (lkysow) [reply agent/consul/leader_peering_test.go:7]: Sounds good. Should we add a linter for this?

File: agent/consul/leader_peering_test.go (hunk 2 above)
Comment (lkysow) [reply agent/consul/leader_peering_test.go:395]: I'm not seeing that method on `peeringServer`

Comment (rboyer) [reply agent/consul/leader_peering_test.go:395]: `peerStreamServer` has a `ConnectedStreams` method that is a pass through.

Comment (lkysow) [reply agent/consul/leader_peering_test.go:395]: Okay thanks. And just for my knowledge why is that preferable vs reaching through `peeringServer`?

**Pull request merged**